From b6e799bbea76d066936d45f0928269ed8c4896f2 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Ville=20Syrj=C3=A4l=C3=A4?= <ville.syrjala@linux.intel.com>
Date: Tue, 19 Dec 2017 21:50:58 +0200
Subject: [PATCH 09/27] matrix: Add weston_matrix_diag()
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Add weston_matrix_diag() which gives us a diagonal matrix
populated with the elements from the passed in vector.

Commit message edit (Harish krupo)

v2: Set matrix type to WESTON_MATRIX_TRANSFORM_OTHER (Pekka)

Signed-off-by: Ville Syrjälä <ville.syrjala@linux.intel.com>
Signed-off-by: Harish Krupo <harish.krupo.kps@intel.com>
Signed-off-by: alex.chae <alex.chae@intel.com>
---
 include/libweston/matrix.h |  2 ++
 shared/matrix.c            | 13 +++++++++++++
 2 files changed, 15 insertions(+)

diff --git a/include/libweston/matrix.h b/include/libweston/matrix.h
index be4d4eb0..d8306501 100644
--- a/include/libweston/matrix.h
+++ b/include/libweston/matrix.h
@@ -60,6 +60,8 @@ void
 weston_matrix_rotate_xy(struct weston_matrix *matrix, float cos, float sin);
 void
 weston_matrix_transform(struct weston_matrix *matrix, struct weston_vector *v);
+void
+weston_matrix_diag(struct weston_matrix *matrix, const struct weston_vector *v);
 
 int
 weston_matrix_invert(struct weston_matrix *inverse,
diff --git a/shared/matrix.c b/shared/matrix.c
index 4e8d6b40..532355a2 100644
--- a/shared/matrix.c
+++ b/shared/matrix.c
@@ -129,6 +129,19 @@ weston_matrix_transform(struct weston_matrix *matrix, struct weston_vector *v)
 	*v = t;
 }
 
+WL_EXPORT void
+weston_matrix_diag(struct weston_matrix *matrix, const struct weston_vector *v)
+{
+	unsigned c;
+
+	weston_matrix_init(matrix);
+
+	for (c = 0; c < 4; c++)
+		matrix->d[c * 4 + c] = v->f[c];
+
+	matrix->type = WESTON_MATRIX_TRANSFORM_OTHER;
+}
+
 static inline void
 swap_rows(double *a, double *b)
 {
-- 
2.21.3

